File: python/树/865. 具有所有最深节点的最小子树.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        def dfs(root):
            if not root: return 1
            return 1 + max(dfs(root.left),dfs(root.right))
        left,right = dfs(root.left),dfs(root.right)
        if left == right:
            return root
        elif left > right:
            return self.subtreeWithAllDeepest(root.left)
        else:
            return self.subtreeWithAllDeepest(root.right)

File: go/tree/865. 具有所有最深节点的最小子树.go
package tree

func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
	var dfs func(*TreeNode) int
	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		left, right := dfs(root.Left), dfs(root.Right)
		if left > right {
			return left + 1
		} else {
			return right + 1
		}
	}
	left, right := dfs(root.Left), dfs(root.Right)
	if left == right {
		return root
	} else if left > right {
		return subtreeWithAllDeepest(root.Left)
	} else {
		return subtreeWithAllDeepest(root.Right)
	}
}
